Document SunRay on Hipster added (#150)
OpenIndiana/oi-docs: docs/handbook/sunray.md

Limitations:

multihead is possible but the Display Switcher Applet on gnome-panel on the second screen won't start. So you can't reopen closed windows on the second screen.

</div>

## Install

Sun Ray Software is still downloadable at http://edelivery.oracle.com.

install required packages

```shell
# pkg install SUNWut-srss SUNWut-srwc SUNWuti  cde/cde-runtime library/motif isc-dhcp
```

### configure isc-dhcp

ISC DHCP has replaced the Sun DHCP also on Hipster but was already the DHCP server for SunRay on Linux.
So SunRay on Hipster has to use DHCP scripts like on Linux install. But it is possible to configure DHCP manually also.

```
# DHCPDARGS="nge1";

ddns-update-style none;

# option definitions common to all supported networks...
option domain-name "sunray.lan";
option domain-name-servers 192.168.1.6, 192.168.1.1;
option subnet-mask 255.255.255.0;
option broadcast-address 192.168.1.255;
option routers 192.168.1.1;
option space SunRay;
option SunRay.Intf code 33 = text;
option SunRay.Intf "nge1";
option SunRay.AuthSrvr code 21 = ip-address;

> option SunRay.AuthSrvr 192.168.1.2;

Author: https://docs.oracle.com/cd/E25749_01/E25745/html/Alternate-Client-Initialization-Reqs-Using-DHCP.html#Alternate-Vendor-Specific-DHCP-Options

> option SunRay.AuthSrvr 192.168.1.2;

Reviewer: Thanks. Put it into the documentation somewhere, please.

option SunRay.AltAuth code 35 = array of ip-address;
option SunRay.AltAuth 192.168.1.2;
option SunRay.FWSrvr code 31 = ip-address;
option SunRay.FWSrvr 192.168.1.2;
option SunRay.NewTVer code 23 = text;

> option SunRay.NewTVer "4.2_140993-05_2010.08.25.23.16";

Reviewer 2: This is needed, but needs clarification. These parameters define firmware download server and expected versions.

option SunRay.LogHost 192.168.1.2;
option SunRay.LogKern code 25 = integer 8;
option SunRay.LogKern 6;
option SunRay.LogNet code 26 = integer 8;
option SunRay.LogNet 6;
option SunRay.LogUSB code 27 = integer 8;
option SunRay.LogUSB 6;
option SunRay.LogVid code 28 = integer 8;
option SunRay.LogVid 6;
option SunRay.LogAppl code 29 = integer 8;
option SunRay.LogAppl 6;
option SunRay.sunray-servers code 68 = text;
option SunRay.sunray-servers "oi-sr.sunray.lan";
option SunRay.sunray-config-servers code 67 = text;
option SunRay.sunray-config-servers "oi-sr.sunray.lan";

default-lease-time 86400;
max-lease-time 86400;

class "sun" {
  match if substring (option vendor-class-identifier, 0, 4) = "SUNW";
}

log-facility local7;

host sr1.sunray.lan {
  hardware ethernet 00:14:4f:57:a0:c1;
  fixed-address 192.168.1.50;
  vendor-option-space SunRay;
}

subnet 192.168.1.0 netmask 255.255.255.0 {
  range 192.168.1.51 192.168.1.59;
}
```

this config file has to linked to `/etc/inet/dhcp4.conf`

For SunRay Software switch to isc-dhcp

```
root@oi-sr:/etc/opt/SUNWut# ln -s /opt/SUNWut/lib/dhcp/isc dhcp
```

`utadm` command need some updates shown in this patch:

> ```

Reviewer: Does "```patch" make a difference here?

Author: yes looks better

+       #     return 1
+       # fi
        H=`getIfname ${INTF}`
        if [ -z "${H}" ]; then
            print -u2 "Error: interface ${INTF} is partially configured.  Hostname not" \
@@ -3453,19 +3455,7 @@ function getIfname {
        awk '{print $1 ; exit}' "${HOSTNAME_R}.$1"
        return
     else
-       # With ipadm, there's no way to create a persistent
-       # configuration without it immediately being reflected in the
-       # runtime configuration, nor does ipadm provide any convenient
-       # way to find the IPv4 addrobj for the interface, so just
-       # check runtime configuration via ifconfig.
-       IPADDR=$(ifconfig $1 | awk '
-               /[      ]inet /{
-                       for (i = 1; i < NF; ++i) {
-                               if ($i == "inet") {
-                                       print $(i+1)
-                               }
-                       }
-               }')
+       IPADDR=`ipadm show-addr -p -o ADDR $1/ | awk '{split($1,a,"/"); print a[1] }'`
        if [ -z "$IPADDR" ]; then
            print -u2 "Error in parsing of ifconfig addresses"
            exit 2
```

### JRE

> The SunRay Software needs the Sun/Oracle JRE 1.7

Reviewer: Please, put period (`.`) here at the end of the sentence and at other places too.

Can we use our `runtime/java/openjdk7`? If not link to a usable Oracle JRE would be helpful.

## Config

You can run `utconfig` and `utadm` run in common manner

### GDM

> the gdm service has to be enabled

Reviewer: `s/gdm/GDM/` please.

```

> # svcadm enable graphical-login/gdm

Reviewer: It seems that GDM is not available at this point, move the section further below.

```

# Revive GNOME on current Hipster

Be sure you have installed

```shell
# pkg install libwnck
# pkg install libbonoboui

> # pkg install gnome-themes-standard

Reviewer: Make it `pkg install libwnck libbonoboui gnome-themes-standard`.

```

these must not be empty metapackages or obsoleted packages.

> ## update older Hipster

Reviewer: Capitalize the section title here and at other places.

If you have SunRay running on Hipster with GNOME, you can prevent remove GNOME with freeze the GNOME packages and release the version locks

```shell
# pkg freeze gdm gnome-session gnome-panel metacity libgnomekbd gnome-settings-daemon gweather

# pkg change-facet facet.version-lock.library/desktop/gnome-desktop=false
# pkg change-facet facet.version-lock.library/desktop/gnome-menus=false
# pkg change-facet facet.version-lock.gnome/gnome-session=false
# pkg change-facet facet.version-lock.system/display-manager/gdm=false
# pkg change-facet version-lock.library/desktop/libgweather=false
# pkg change-facet version-lock.gnome/window-manager/metacity=false
```

## install from current Hipster (2019)

If you install new Hipster it is necessary to have the GNOME packages before (or after) obsoleting

The last know version of OI GNOME packages are:

    libgweather@2.30.3
    gnome-session@2.32.1
    gnome-panel@2.32.1
    metacity@2.30.3
    gdm@2.30.7
    libgnomekbd@2.30.0
    gnome-settings-daemon@2.32.1

These are still in hipster repo, but are obsoleted empty metapackages.

> [here](http://pkg.toc.de/sunray/)  can you found the old GNOME packages with newer release date, so that these should be able to install it on current Hipster.

Reviewer: "here can you found the" -> "Here you can find".

Add a following sentence (or a similar one): This publisher is provided and supported by community member Carsten Grzemba not by the OpenIndiana project.

> For that add the publisher in that kind, that it can override packages of publisher openindiana.org

Reviewer: ```suggestion
For this to work add this publisher in a way, that it takes precedence over the default openindiana.org publisher:
```

```shell
# pkg set-publisher --search-before=openindiana.org -g http://pkg.toc.de/sunray sunray
# pkg set-publisher --non-sticky openindiana.org
```

> and install the all packages from publisher sunray.

Reviewer: ```suggestion
and install all packages from sunray publisher.
```

Reviewer: Please, also show **how** one does that.

> **The version locks have to release anyway.**

Reviewer: I don't understand this. Please describe **why** and **how** locks are released.

Author: I made some changes to clarify this.

> #### Xscreensaver

Reviewer: Make it [XScreenSaver](https://www.jwz.org/xscreensaver/) here and other places.

> latest Hipster delivers Xscreensaver only in 64bit. The PAM module of SunRay are only shipped in 32bit.

Reviewer: ```suggestion
Latest Hipster delivers XScreenSaver only in 64-bit. SunRay PAM modules are shipped as 32-bit-only.
```

Author: xscreensaver delivers only bins and no so-libs, so the build bit size should not be of interest. I don't know anymore why I belived it is necessary. I remove this section.

> So we need a Xscreensaver package with 32bit bins (recompile with enabled 32bit or take it from [here](http://pkg.toc.de/sunray) ).

Reviewer: The user was asked to install all packages from the sunray publisher, so they already have your 32-bit XScreenSaver. Show the user how to install XScreenSaver from your repo directly (drop the compile yourself note), or drop the XScreenSaver section completely.

# Login Screen (gdm-greeter) won't reapear after logout

> If after logout the gdm-login not reappears, helps the following script installed at `/opt/SUNWut/lib/gdm/revivesrsession.py` and called by the helper script `/etc/opt/SUNWut/gdm/SunRayPostSession/helpers/revivesession`

Reviewer: ```suggestion
After logout the gdm-login won't reappear. Install the following script as `/opt/SUNWut/lib/gdm/revivesrsession.py` and call it by the `/etc/opt/SUNWut/gdm/SunRayPostSession/helpers/revivesession` helper script:
```

```
/opt/SUNWut/lib/gdm/revivesrsession.py &
```

```python
#!/usr/bin/env python

'''
 Reconnect SunRay X-session: some times after logout, the gdm will not start login screen on the X-session and the DTU remains in state 26D
 INSTALL: - cp to /opt/SUNWut/lib/gdm/revivesrsession.py
          - create script /etc/opt/SUNWut/gdm/SunRayPostSession/helpers/revivesession:
            /opt/SUNWut/lib/gdm/revivesrsession.py &
'''

import subprocess as sp
import re
import pdb
import time
import logging

logformat = "%(asctime)s %(levelname)s:%(message)s"
logging.basicConfig(format = logformat)
logger = logging.getLogger(__name__)

logger.setLevel(logging.DEBUG)

time.sleep(10)

pid = sp.Popen(['pgrep','gdm-binary'], stdout=sp.PIPE).stdout.readline().strip()
logger.debug("GDM pid %s ", pid)
dpl = [ p.split()[2].strip(':')
    for p in sp.Popen(['ptree', pid], stdout=sp.PIPE).stdout.readlines() if '/opt/SUNWut/lib/Xnewt' in p ]
logger.debug("Xnewt pid {}".format( dpl))

# error -4 gdm-simple-slave not started for Display, no UT sessions

> for sess in sp.Popen(['pfexec','/opt/SUNWut/sbin/utsession','-px'], stdout=sp.PIPE).stdout.readlines():

Reviewer: Does one need to have `pfexec` setup first? If so what privileges?

Author: You are right. My first thought was it is running by user gdm, but it is running as root. So it is not necessary and will remove pfexec

<!--

The contents of this Documentation are subject to the Public Documentation License Version 1.01
 (the "License"); you may only use this Documentation if you comply with the terms of this License.
A copy of the License is available at http://illumos.org/license/PDL.

The Original Documentation is _________________.

The Initial Writer of the Original Documentation is Carsten Grzemba Copyright (C) 2020.
All Rights Reserved. (Initial Writer contact(s): cgrzemba@opencsw.org).

Contributor(s): ______________________________________.

Portions created by ______ are Copyright (C)_________[Insert year(s)].
All Rights Reserved. (Contributor contact(s):________________[Insert hyperlink/alias]).

-->

> # Sun Ray Server on Openindiana Hipster

Reviewer: ```suggestion
# Sun Ray Server Software on OpenIndiana Hipster
```

> Some notes for installation Sun Ray Server on Openindiana Hipster.

Reviewer: ```suggestion
Some notes for installation of Sun Ray Server Software on OpenIndiana Hipster.
```

> Some notes for installation Sun Ray Server on Openindiana Hipster.

Reviewer: ```suggestion
Some notes for installation Sun Ray Server Software on OpenIndiana Hipster.
```

<i class="fa fa-info-circle fa-lg" aria-hidden="true"></i> **NOTE:**
<div class="well">
Since GNOME is replaced by Mate in Hipster, installation is a little bit more difficult. For Sun Ray it is still necessary to use GNOME GDM and some other GNOME applications. But it is possible to use also Mate applications like pluma or atril.

Limitations:

> multihead is possible but the Display Switcher Applet on gnome-panel on the second screen won't start. So you can't reopen minimized  windows on the second screen.

Reviewer: ```suggestion
Multihead is possible but the Display Switcher Applet on GNOME Panel on the second screen won't start. Thus minimized windows on the second screen can't be reopened.
```

</div>

## Install

> Sun Ray Software is still downloadable at http://edelivery.oracle.com.

Reviewer: ```suggestion
Sun Ray Server Software still can be downloaded at http://edelivery.oracle.com (registration required).
```

> install IPS packages of the downloaded Sun Ray Server Software

Reviewer: ```suggestion
Install IPS packages of the downloaded Sun Ray Server Software:
```

```shell
# pkg install SUNWut-srss SUNWut-srwc SUNWuti  cde/cde-runtime library/motif isc-dhcp
```

> Caused by package dependencies these packages are installed finally:

Reviewer: ```suggestion
Due to package dependencies following packages will be installed:
```

SUNWutscr (sunray)                                4.5.4.0.38-0.0             i--
SUNWutsrs (sunray)                                4.5.5.0.38-0.0             i--
SUNWutstk (sunray)                                4.5.0.0.44-0.0             i--
SUNWutsto (sunray)                                4.5.0.0.44-0.0             i--
SUNWutstr (sunray)                                4.5.0.0.44-0.0             i--
SUNWutsvt (sunray)                                1.1.0.0.3-0.0              i--
SUNWuttsc (sunray)                                2.6.4.0.11-0.0             i--
SUNWuttscd (sunray)                               2.6.4.0.11-0.0             i--
SUNWuttscr (sunray)                               2.6.4.0.11-0.0             i--
SUNWutu (sunray)                                  4.5.0.0.44-0.0             i--
SUNWutwa (sunray)                                 4.5.2.0.5-0.0              i--
SUNWutwar (sunray)                                4.5.0.0.44-0.0             i--
SUNWutwh (sunray)                                 4.5.0.0.44-0.0             i--
SUNWutwl (sunray)                                 4.5.0.0.44-0.0             i--
```

> ### configure isc-dhcp

Reviewer: ```suggestion
### Configure isc-dhcp
```

> ISC DHCP has replaced the Sun DHCP server also on Hipster but was already the DHCP server for Sun Ray Server on Linux.

Reviewer: ```suggestion
On Hipster ISC DHCP replaced the Sun DHCP server, but that's not a problem as ISC DHCP server for Sun Ray Server was used on Linux already.
```

> So Sun Ray Server on Hipster has to use DHCP scripts like on Linux install. But it is possible to configure DHCP manually also and skipping the step of `utadm` . Here an example:

Reviewer: ```suggestion
So Sun Ray Server on Hipster has to use DHCP scripts like on Linux. But it is also possible to configure DHCP manually and to skip the `utadm` step. Here an example:
```

option SunRay.FWSrvr 192.168.1.2;
option SunRay.NewTVer code 23 = text;
option SunRay.NewTVer "11.1.7.0_38_2016.10.27.13.09";
option SunRay.AuthPort code 22 = integer 16;
option SunRay.AuthPort 7009;
option SunRay.LogHost  code 24 = ip-address;
option SunRay.LogHost 192.168.1.2;
option SunRay.LogKern code 25 = integer 8;
option SunRay.LogKern 6;
option SunRay.LogNet code 26 = integer 8;
option SunRay.LogNet 6;
option SunRay.LogUSB code 27 = integer 8;
option SunRay.LogUSB 6;
option SunRay.LogVid code 28 = integer 8;
option SunRay.LogVid 6;
option SunRay.LogAppl code 29 = integer 8;
option SunRay.LogAppl 6;
option SunRay.sunray-servers code 68 = text;
option SunRay.sunray-servers "oi-sr.sunray.lan";
option SunRay.sunray-config-servers code 67 = text;
option SunRay.sunray-config-servers "oi-sr.sunray.lan";

default-lease-time 86400;
max-lease-time 86400;

class "sun" {
  match if substring (option vendor-class-identifier, 0, 4) = "SUNW";
}

log-facility local7;

host sr1.sunray.lan {
  hardware ethernet 00:14:4f:57:a0:c1;
  fixed-address 192.168.1.50;
  vendor-option-space SunRay;
}

subnet 192.168.1.0 netmask 255.255.255.0 {
  range 192.168.1.51 192.168.1.59;
}
```

> The vendor specific DHCP options described at [Sun Ray Software: Alternate Client Initialization Reqs Using DHCP](https://docs.oracle.com/cd/E25749_01/E25745/html/Alternate-Client-Initialization-Reqs-Using-DHCP.html#Alternate-Vendor-Specific-DHCP-Options).

Reviewer: ```suggestion
Vendor specific DHCP options are described in the [Sun Ray Software: Alternate Client Initialization Reqs Using DHCP](https://docs.oracle.com/cd/E25749_01/E25745/html/Alternate-Client-Initialization-Reqs-Using-DHCP.html#Alternate-Vendor-Specific-DHCP-Options) document.
```

> The config file `/etc/dhcp/dhcpd.conf` has to link to `/etc/inet/dhcp4.conf` for use with the `svc:/network/dhcp/server:ipv4` service.

Reviewer: ```suggestion
The config file `/etc/dhcp/dhcpd.conf` has to link to `/etc/inet/dhcp4.conf` so that the `svc:/network/dhcp/server:ipv4` service can leverage it.
```

> For Sun Ray Software switch to isc-dhcp:

Reviewer: ```suggestion
To switch Sun Ray Software to ISC DHCP run:
```

```
root@oi-sr:/etc/opt/SUNWut# ln -s /opt/SUNWut/lib/dhcp/isc dhcp
```

> `utadm` command need some updates shown in this patch:

Reviewer: ```suggestion
`utadm` command needs changes shown in the following patch:
```

+       #     return 1
+       # fi
        H=`getIfname ${INTF}`
        if [ -z "${H}" ]; then
            print -u2 "Error: interface ${INTF} is partially configured.  Hostname not" \
@@ -3453,19 +3455,7 @@ function getIfname {
        awk '{print $1 ; exit}' "${HOSTNAME_R}.$1"
        return
     else
-       # With ipadm, there's no way to create a persistent
-       # configuration without it immediately being reflected in the
-       # runtime configuration, nor does ipadm provide any convenient
-       # way to find the IPv4 addrobj for the interface, so just
-       # check runtime configuration via ifconfig.
-       IPADDR=$(ifconfig $1 | awk '
-               /[      ]inet /{
-                       for (i = 1; i < NF; ++i) {
-                               if ($i == "inet") {
-                                       print $(i+1)
-                               }
-                       }
-               }')
+       IPADDR=`ipadm show-addr -p -o ADDR $1/ | awk '{split($1,a,"/"); print a[1] }'`
        if [ -z "$IPADDR" ]; then
            print -u2 "Error in parsing of ifconfig addresses"
            exit 2
```

> ### JRE

Reviewer: ```suggestion
### Java Runtime Environment
```

But why is this section here? Does the user need to do something about the JRE?

The Sun Ray Software needs the Sun/Oracle JRE 1.7 which is shipped with the Sun Ray Software.

## Config

> You can run `utconfig` and `utadm` run in common manner described in [Sun Ray Software](https://docs.oracle.com/cd/E35310_01/index.html).

Reviewer: ```suggestion
You can run `utconfig` and `utadm` the way described in the [Sun Ray Software](https://docs.oracle.com/cd/E35310_01/index.html) document.
```

> # Use GNOME on current Hipster

Reviewer: ```suggestion
# How to get GNOME on current Hipster
```

> As already mentioned Sun Ray Server cannot handled by lightdm and so we still have to use GDM and GNOME.

Reviewer: ```suggestion
As already mentioned the Sun Ray Server cannot handled by the LightDM and we still have to use GDM and some GNOME components.
```

Be sure you have installed:

```shell
# pkg install libwnck
# pkg install libbonoboui
# pkg install gnome-themes-standard
```

> these must not be empty metapackages or obsoleted packages.

Reviewer: ```suggestion
Make sure that they actually install as in the future they may be eventually obsoleted from Hipster.
```

## update older Hipster

> If you have Sun Ray running on Hipster with GNOME, you can prevent remove GNOME with freeze the GNOME packages and release the version locks

Reviewer: ```suggestion
If you have Sun Ray running on Hipster with GNOME, you can prevent GNOME components removal by "freezing" them and releasing their version locks:
```

```shell
# pkg freeze gdm gnome-session gnome-panel metacity libgnomekbd gnome-settings-daemon gweather

# pkg change-facet facet.version-lock.library/desktop/gnome-desktop=false
# pkg change-facet facet.version-lock.library/desktop/gnome-menus=false
# pkg change-facet facet.version-lock.gnome/gnome-session=false
# pkg change-facet facet.version-lock.system/display-manager/gdm=false
# pkg change-facet version-lock.library/desktop/libgweather=false
# pkg change-facet version-lock.gnome/window-manager/metacity=false
```

> ## install from current Hipster (2019)

Reviewer: ```suggestion
## Install from current Hipster (2019)
```

> If you install new Hipster it is necessary to have the GNOME packages before (or after) obsoleting

Reviewer: ```suggestion
If you installed new Hipster, it is necessary to install GNOME packages before they were obsoleted from Hipster.
```
I am not sure what this sentence tries to say.

> The last know version of OI GNOME packages are:

Reviewer: ```suggestion
Last know working OpenIndiana GNOME packages versions are the following:
```

    libgweather@2.30.3
    gnome-session@2.32.1
    gnome-panel@2.32.1
    metacity@2.30.3
    gdm@2.30.7
    libgnomekbd@2.30.0

>     gnome-settings-daemon@2.32.1

Reviewer: Put it to "```"-block.

These are still in hipster repo, but are obsoleted empty metapackages.

At <http://pkg.toc.de/sunray/> can you found the old GNOME packages with newer release date, so that these should be able to install it on current Hipster.
For that add the publisher in that kind, that it can override packages of publisher openindiana.org

```shell
# pkg set-publisher --search-before=openindiana.org -g http://pkg.toc.de/sunray sunray
# pkg set-publisher --non-sticky openindiana.org
```

and install the all packages from publisher sunray.

**The version locks have to release anyway.**

#### XScreenSaver

Latest Hipster delivers XScreenSaver only in 64-bit. The PAM module of Sun Ray are only shipped as 32-bit-only.

> Thats why we need the XScreenSaver package with 32-bit bins from <http://pkg.toc.de/sunray/>.

Reviewer: ```suggestion
That's why we need the XScreenSaver package with 32-bit bins from <http://pkg.toc.de/sunray/>.
```

## GDM

> the gdm service has to be enabled. lightdm should not run.

Reviewer: ```suggestion
The gdm service has to be enabled. LightDM should not run.
```

```
# svcadm enable graphical-login/gdm

> ```

Reviewer: Also show how to disable LightDM service.

## Install

Sun Ray Software is still downloadable at http://edelivery.oracle.com. Download the Software for Solaris x86-64 and for Linux x86-64. From the Linux software we take scripts for configure ISC DHCP server described later.

Extract the archives.

> The Solaris packages require Sun DHCP server, which we not provide anymore. To solve this problem we update the IPS packages SUNWutr and SUNWuto with the following script:

Reviewer 2: do not provide
